tests/callbacks/test_early_stopping.py

```diff
@@ -344,3 +345,51 @@ def validation_epoch_end(self, outputs):
 def test_early_stopping_mode_options():
     with pytest.raises(MisconfigurationException, match="`mode` can be .* got unknown_option"):
         EarlyStopping(mode="unknown_option")
+
+
+class EarlyStoppingModel(BoringModel):
+
+    def __init__(self, expected_end_epoch):
+        super().__init__()
+        self.expected_end_epoch = expected_end_epoch
+
+    def validation_epoch_end(self, outputs):
+        losses = [8, 4, 2, 3, 4, 5, 8, 10]
+        val_loss = losses[self.current_epoch]
+        self.log('abc', torch.tensor(val_loss))
+        self.log('cba', torch.tensor(0))
+
+    def on_train_end(self) -> None:
+        assert self.trainer.current_epoch == self.expected_end_epoch, 'Early Stopping Failed'
```

Comment:
I would drop this and rather check in the test trainer epoch is as expected, so there is not random inference

Reply:
That's what I originally did but because of how DDP Spawn works, the local trainer's current epoch doesn't seem to be kept in sync which is fair (since it's only kept in sync during trainer on the processes). This is why I had to move it to `on_train_end` because this happens within the spawn process!

Reply:
so can we have both?

Reply:
We could but I'd need to separate out the tests, I don't think its really worth it because it would be a lot of duplication